lib/AST/CMakeLists.txt

```diff
@@ -3,6 +3,57 @@ if (SWIFT_FORCE_OPTIMIZED_TYPECHECKER)
   set(EXTRA_AST_FLAGS "FORCE_BUILD_OPTIMIZED")
 endif()
 
+set(SWIFTAST_LINK_LIBRARIES
```

Comment:
This really makes it difficult to follow with all the appending.  Can you please switch this to "modern" CMake instead - use a series of `target_link_libraries` instead after the target has been created.

Reply:
I started doing that, and I thought that the only difference between `LINK_LIBRARIES` and `INTERFACE_LINK_LIBRARIES` is whether `target_link_libraries` is called with `PRIVATE` or `INTERFACE`, but when I add
```
  target_link_libraries(swiftAST PRIVATE
    clangBasic
  )
```
I see that `clangBasic` ends up bringing along the LLVM libraries that I wanted to avoid, while it did not bring them along when I passed it with `LINK_LIBRARIES` to `add_swift_host_library`.

Any hints for me, I find dealing with this CMake black magic, quite frustrating 😞

Reply:
Linking against `clangBasic` will of course bring in the dependencies for the library.  However, since this will be linked statically, all of the unused functions should be dropped.  Or is the intention something else?

Reply:
The intention is to not even have to build the additional libraries that it brings, I only want it to build `clangBasic`. It seems I can get this behavior if I pass `clangBasic` as a member of `LINK_LIBRARIES`.

Reply:
Ok, I see the difference is that `add_swift_host_library` does some processing of `LINK_LIBRARIES` and passes them to `target_link_libraries` with an absolute path. Code is here: https://github.com/apple/swift/blob/cb09c33a6b58f04fba4125410332b31091f7ffc7/cmake/modules/AddSwift.cmake#L1096

I need to pass the list as `LINK_LIBRARIES`. And I might as well do the same for `INTERFACE_LINK_LIBRARIES` and pass them via `add_swift_host_library` because it's better to follow what other libraries are doing and who knows what else may be happening with that list.

What do you think, do you have a suggestion to improve the readability here but still pass the lists to `add_swift_host_library` ?

Reply:
@compnerd ?

Reply:
No, I don't think that there is a way to accomplish both.  Modern CMake avoids this and tends to prefer to just append via `target_link_libraries`.  If you need to do that for that specific library, you could do something evil like use a generator expression: `$<TARGET_PROPERTY:clangBasic,IMPORTED_LOCATION>` or maybe `$<TARGET_LINKER_FILE:clangBasic>`.

Reply:
> Modern CMake avoids this and tends to prefer to just append via `target_link_libraries`

Ok but It's not clear to me why this is important to do here for `swiftAST` as part of this patch, while every other library, that the swift project defines, is defining its dependencies by passing them to `add_swift_host_library`.

Reply:
Note that I can't just use `target_link_libraries`, I would also have to duplicate what that piece of code I linked to earlier is doing (which is part of `add_swift_host_library`).

Reply:
Well, I am slowly moving them towards the modern CMake mechanisms, which just makes everything more readable as well as easier to debug.  The others will follow suit soon too.

Reply:
Please note above comment:

> I can't just use `target_link_libraries`, I would also have to duplicate what that piece of code I linked to earlier is doing (which is part of `add_swift_host_library`)

`add_swift_host_library` is doing more things than just passing the list of libraries to `target_link_libraries`.

Reply:
@akyrtzi - oh, did the `$<TARGET_PROPERTY:clangBasic,IMPORTED_LOCATION>` or `$<TARGET_LINKER_FILE:clangBasic>` not do what you needed?

Reply:
To clarify, you are suggesting to avoid passing the dependencies to `add_swift_host_library` but add them via `target_link_libraries`. That would happen not just when `SWIFT_BUILD_ONLY_SYNTAXPARSERLIB` is enabled but when it is not enabled as well.

What I'm suggesting is that I'm not comfortable making such a change for the common case `SWIFT_BUILD_ONLY_SYNTAXPARSERLIB=FALSE` because `add_swift_host_library` is doing more complex things to the list of dependencies than just pass them to `target_link_libraries`.

Also it's not clear to me what `$<TARGET_PROPERTY:clangBasic,IMPORTED_LOCATION>` is doing and where exactly it needs to be inserted in the CMake file.

Reply:
Okay, I suppose I can do that in a follow up.  And, yes, I was asking that we always use the `target_link_libraries` approach.

Reply:
Please see updated patch.

Reply:
I'm going to assume that changes after your https://github.com/apple/swift/pull/22831 are close to your satisfaction, but if you have additional feedback I'll be happy to do a follow-up PR.